### PR TITLE
Updating to ST4 with many changes

### PR DESCRIPTION
**Pull request by TerminalFi**
### What

- Updating to support various features of ST4. 
- Dropping support for ST3
- Adding a dedicate Regex Search using Input Panel and Live Previewing

### Note

Worth noting this would require a update to Package Control to continue to allow the previous release for ST3 and the new releases targeting ST4

<details>
<summary>Video Example</summary>

https://user-images.githubusercontent.com/32599364/171492361-93e56e9a-5a45-4d45-9809-555ffadbd57a.mp4

</details>

### CODE REVIEW

File: Main.sublime-menu [hunk 1]
```diff
@@ -1,108 +1,103 @@
...
+    "id": "goto",
+    "children": [
+      {
+        "caption": "Jump to …",
+        "command": "jump_to_interactive"
+      }
+    ]
+  },
   {
     "caption": "Preferences",
```
Comment (jrappen) [inline Main.sublime-menu:63]: please also remove overrides for caption and mnemonic, simply targetting by id for existing menu tree entries should suffice

File: Main.sublime-menu [hunk 2]
```diff
@@ -1,108 +1,103 @@
...
     "mnemonic": "n",
     "id": "preferences",
-    "children":
-    [
+    "children": [
       {
         "caption": "Package Settings",
```
Comment (jrappen) [inline Main.sublime-menu:68]: same as above

**Comment by jrappen** [comment]
since you moved to py3.8 you could also make use of f-strings and/or `typing`

**Comment by TheSecEng** [comment]
@jrappen I did the majority of typings. I am not super strong in understanding how to write Generator typings. Could use some guidance

File: selection_fields.py [hunk 3]
```diff
@@ -6,46 +6,46 @@
 
 def get_settings(key, default=None):
     """Get the setting specified by the key."""
-    settings = sublime.load_settings("MultiEditUtils.sublime-settings")
+    settings = sublime.load_settings('MultiEditUtils.sublime-settings')
     return settings.get(key, default)
 
 
 def _get_settings(key, default=None):
     """
```
Comment (jrappen) [inline selection_fields.py]: you skipped changing the quotes styles in this line

Comment (jrappen) [reply selection_fields.py]: disregard, was fixed in the next commit

File: MultiEditUtils.py [hunk 4]
```diff
@@ -474,7 +470,7 @@ def splitByCase(self, aString):
 
         return newGroups
 
-    def analyzeCase(self, aString):
+    def analyzeCase(self, value) -> int:
```
[the remark this answers is not in the document]
Comment (TheSecEng) [reply MultiEditUtils.py:473]: This gives Type annotation illegal due to variable use

**Comment by AllanLRH** [comment]
I still use ST3, as I haven't bought a license for ST4 (I mostly use VS Code).
If breaking backward compatability with ST3, is it possible to release the ST3-compatible version under another name, or perhaps the ST4 under another name?

**Comment by TheSecEng** [comment]
> I still use ST3, as I haven't bought a license for ST4 (I mostly use VS Code).
> If breaking backward compatability with ST3, is it possible to release the ST3-compatible version under another name, or perhaps the ST4 under another name?

This would be the plan. To create a new release specific to ST 4. Which would preserve st 3 compatibility

File: MultiEditUtils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheSecEng) [reply MultiEditUtils.py:473]: @jrappen thoughts?

**Comment by philippotto** [comment]
Hi @TheSecEng,

sorry for the late reply. I don't get notifications in repositories unless I'm directly mentioned.

First of all, thank you for your work and for opening a PR. Unfortunately, the PR is relatively big which makes it time-consuming to review. Splitting the PR into multiple smaller ones would help a lot. Especially, the changes of formatting/code style (i.e., double quotes to single quotes) blows up the line changes. So, please split this up into a separate PR or avoid it altogether, as I don't see an immediate value for that.

Furthermore, I have some questions:

> Updating to support various features of ST4.

I don't really understand this point. This package is already compatible with ST4 (I'm using it myself for ST4). Are there any problems with it? Also, the package is already listed as ST4 compatible on packagecontrol.io.

> Dropping support for ST3

Why is this necessary? Can we avoid this?

> Adding a dedicate Regex Search using Input Panel and Live Previewing

Sounds interesting! I don't really use the regex-related features of MEU myself (they were added by contributors). However, if you split that feature into a separate PR, it would be easier to review it and judge whether it makes sense to add that feature.

> I did the majority of typings.

Are the typings validated by mypy or something else? Without that, it doesn't really make sense to add them in my opinion. Also, this should be a separate PR.

Thanks again for your contribution!

**Comment by TheSecEng** [comment]
Sure I can close this and open smaller prs. I really did they updates on my private copy and figured I’d see if anyone was interested. 

As for the st3 compatibility. I can retain it, but it just adds a lot of work which I don’t see value, we can just create a st3 branch and maintain st3 a current state and then new features can be focused on for ST 4 

as some feature this uses aren’t available in ST 3

**Comment by jrappen** [comment]
I'd assume a very high majority of users are on ST4 already. ST3 support is simply a waste of time & resources.

File: MultiEditUtils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrappen) [reply MultiEditUtils.py:473]: I'd have to check. No time right now, maybe tomorrow. Also on mobile at the moment.

[the remark this answers is not in the document]
Comment (FichteFoll) [reply MultiEditUtils.py:473]: `Case` is defined as an instance of the `Case` namedtuple, which is … quite an odd usage of it.

https://github.com/TheSecEng/Sublime-MultiEditUtils/blob/7587f24f64243825d4ebaebe62313e627a2382b4/MultiEditUtils.py#L391

A better solution would be [an enum](https://docs.python.org/3/library/enum.html).

**Pull request closed without merging**